test/queue.f90

```diff
@@ -0,0 +1,171 @@
+
+!
+! Define the subroutines and functions
+!
+contains
+
+! queue_create --
+!     Create and initialise a queue
+! Arguments:
+!     queue      Pointer to new queue
+!     capacity   The number of data that can be stored
+! Note:
+!     This version assumes a shallow copy is enough
+!     (that is, there are no pointers within the data
+!     to be stored)
+!     It also assumes the argument queue does not already
+!     refer to a queue. Use queue_destroy first to
+!     destroy an old queue.
+!
+!     There is no check that the capacity is positive!
+!
+subroutine queue_create( queue, capacity )
+type(QUEUE_STRUCT), pointer        :: queue
+integer                            :: capacity
+
+allocate( queue )
+allocate( queue%data(1:capacity) )
+
+queue%full  = .false.
+queue%start = 1
+queue%end   = 0
+end subroutine queue_create
+
+! queue_destroy --
+!     Destroy a queue
+! Arguments:
+!     queue       Pointer to the queue to be destroyed
+! Note:
+!     This version assumes that there are no
+!     pointers within the data that need deallocation
+!
+subroutine queue_destroy( queue )
+type(QUEUE_STRUCT), pointer  :: queue
+
+deallocate( queue%data )
+deallocate( queue )
+end subroutine queue_destroy
+
+! queue_empty --
+!     Check if the queue is empty
+! Arguments:
+!     queue       Pointer to the queue
+! Result:
+!     logical indicating if the queue is
+!     empty or not
+!
+logical function queue_empty( queue )
+type(QUEUE_STRUCT), intent(in)  :: queue
+
+queue_empty = .not. queue%full .and. &
+    queue%end .eq. queue%start - 1
+
+end function queue_empty
+
+! queue_full --
+!     Check if the queue is full
+! Arguments:
+!     queue       Pointer to the queue
+! Result:
+!     logical indicating if the queue is
+!     full or not
+!
+logical function queue_full( queue )
+type(QUEUE_STRUCT), intent(in)  :: queue
+
+queue_full = queue%full
+
+end function queue_full
+
+! queue_start_data
+!     Return the data stored at the start,
+!     but leave them in
+! Arguments:
+!     queue      Queue to be examined
+! Result:
+!     Data stored at the start
+! Note:
+!     With an empty queue, random data
+!     are returned!
+!
+function queue_start_data( queue ) result(data)
+type(QUEUE_STRUCT), intent(in) :: queue
+integer                        :: data
+
+data = queue%data(queue%start)
+
+end function queue_start_data
+
+! queue_retrieve_data
+!     Return the data stored at the top,
+!     and remove them from the queue
+! Arguments:
+!     queue      Queue to be examined
+! Result:
+!     Data stored at the top, afterwards
+!     removed
+! Note:
+!     With an empty queue, random data
+!     are returned!
+!
+function queue_retrieve_data( queue ) result(data)
+type(QUEUE_STRUCT)             :: queue
+integer                        :: data
+
+data = queue%data(queue%start)
+
+if ( .not. queue_empty(queue) ) then
+    queue%start = queue%start + 1
+    if ( queue%start .gt. size(queue%data) ) then
+        queue%start = 1
```

Comment:
May be make queue full

Reply:
When this condition is reached the queue can not be full since we are removing an element. The reason that is check exists is because the queue is circular. Meaning when we add and remove from the array, they aren't being added to the beginning or end. So this check allows the queue to wrap around and make sure that the full array is utilized.